File: src/iftun/iftun.c
#include "iftun.h"

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/if.h>
#include <linux/if_tun.h>

int iftun_alloc(char devname[16], int flags)
{
    struct ifreq ifr;

    int fd = open("/dev/net/tun", O_RDWR);
    if (fd < 0) {
        perror("open(/dev/net/tun)");
        return -1;
    }

    memset(&ifr, 0, sizeof(ifr));
    ifr.ifr_flags = flags; // ex: IFF_TUN | IFF_NO_PI

    if (devname && devname[0] != '\0') {
        strncpy(ifr.ifr_name, devname, IFNAMSIZ - 1);
        ifr.ifr_name[IFNAMSIZ - 1] = '\0';
    }

    if (ioctl(fd, TUNSETIFF, (void*)&ifr) < 0) {
        perror("ioctl(TUNSETIFF)");
        close(fd);
        return -1;
    }

    strncpy(devname, ifr.ifr_name, 16 - 1);
    devname[16 - 1] = '\0';
    return fd;
}

int iftun_copy_forever(int src_fd, int dst_fd)
{
    unsigned char buf[2000];
    for (;;) {
        ssize_t n = read(src_fd, buf, sizeof(buf));
        if (n < 0) {
            perror("read(tun)");
            return -1;
        }
        ssize_t off = 0;
        while (off < n) {
            ssize_t w = write(dst_fd, buf + off, (size_t)(n - off));
            if (w < 0) {
                perror("write(dst)");
                return -1;
            }
            off += w;
        }
    }
}

File: src/extremite/iftun.py
import os
import fcntl
import struct
import subprocess

# Constants from linux/if_tun.h
TUNSETIFF = 0x400454ca
IFF_TUN   = 0x0001
IFF_NO_PI = 0x1000 # Warning: If you remove this, you get extra 4 bytes header!

def tun_alloc(dev_name="tun0"):
    """
    Creates a TUN interface.
    Returns: file descriptor (int) for the interface.
    """
    # Open the clone device
    try:
        tun_fd = os.open("/dev/net/tun", os.O_RDWR)
    except FileNotFoundError:
        print("Error: /dev/net/tun not found. Are you on Linux?")
        exit(1)
    except PermissionError:
        print("Error: Permission denied. Did you run with sudo?")
        exit(1)

    # Prepare the struct ifreq (interface request)
    # Format '16sH': 16 bytes for name, 2 bytes (Short) for flags
    # We strip the name to 15 chars + null terminator
    ifr_name = dev_name.encode('utf-8')[:15]
    ifr_flags = IFF_TUN | IFF_NO_PI
    
    # Pack the data into binary format expected by C ioctl
    ifr = struct.pack('16sH', ifr_name, ifr_flags)
    
    # Send the ioctl command to the kernel
    fcntl.ioctl(tun_fd, TUNSETIFF, ifr)
    
    print(f"Interface {dev_name} created successfully.")
    return tun_fd
